### PR TITLE
Enables RLN compilation under CI env var

### PR DESCRIPTION
**Pull request by staheri14**
With this PR, all the code blocks whose compilation are conditioned to the `RLN` flag will get compiled in the GH and Jenkins builds by checking the `CI` environment variable. This will allow rln-relay tests to be included in the CI checks.
The local compilation of rln-related code blocks still depends on the `RLN` flag.

Additionally, due to the issue described in #833 the rln-relay tests that require interaction with ganache-cli are ~commented out~ moved under `onchain_rln` compiler flag until the said issue gets resolved. ~Subsequently, the `ganache-cli` target i.e., `installganache` is also disabled in the Makefile. Note that this does not affect the dogfooding of waku-rln-relay in the off-chain mode.~

### CODE REVIEW

**Comment by jakubgs** [comment]
I've installed `rust` package view Homebrew on all MacOS hosts in CI: https://github.com/status-im/infra-ci/commit/f7091bc6

**Comment by jakubgs** [comment]
The versions are all currently at `1.56.1`:
```
 > a ci-slave-macos -o -a '/usr/local/bin/rustc --version' 
maci7-01.ms-eu-dublin.slave.ci | CHANGED | rc=0 | (stdout) rustc 1.56.1
maci7-02.ms-eu-dublin.slave.ci | CHANGED | rc=0 | (stdout) rustc 1.56.1
maci7-03.ms-eu-dublin.slave.ci | CHANGED | rc=0 | (stdout) rustc 1.56.1
```
But I cannot guarantee that they will stay consistent in the future.

One of these days I need to look at making Linux CI builds for `nim-waku` use Nix.

**Comment by staheri14** [comment]
@jakubgs Thanks a lot for the `rust` installation!

**Comment by jakubgs** [comment]
I added manually a `echo $PATH | tr ':' '\n'` in the beginning of the `Deps` stage using the `Replay` button, and I got this:
```
/c/ProgramData/scoop/apps/openjdk/current/bin
/c/ProgramData/scoop/apps/pypy3/current/Scripts
/c/BuildTools/VC/Tools/MSVC/14.27.29110/bin
/c/BuildTools/MSBuild/Current/Bin
/c/Qt/5.14.2/msvc2017_64/bin
/c/ProgramData/scoop/apps/inno-setup/current
/c/Users/jenkins/bin
/mingw64/bin
/usr/local/bin
/usr/bin
/usr/bin
/mingw64/bin
/usr/bin
/c/Users/jenkins/bin
/c/Users/admin/go/bin
/c/ProgramData/scoop/apps/gcc/current/bin
/c/ProgramData/scoop/apps/openjdk/current/bin
/c/ProgramData/scoop/apps/gsudo/current
/c/ProgramData/scoop/apps/python/current/Scripts
/c/ProgramData/scoop/apps/python/current
/c/ProgramData/scoop/shims
/c/Windows/system32
/c/Windows
/c/Windows/System32/Wbem
/c/Windows/System32/WindowsPowerShell/v1.0
/c/Windows/System32/OpenSSH
/c/ProgramData/GooGet
/c/Program Files/Google/Compute Engine/metadata_scripts
/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin
/c/Program Files/PowerShell/7
/c/Program Files/Google/Compute Engine/sysprep
/c/Program Files/Threat Stack
/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit
/c/Program Files/WireGuard
/c/Users/jenkins/AppData/Local/Microsoft/WindowsApps
/usr/bin/vendor_perl
/usr/bin/core_perl
```
Bit this is what I get when I check where `cargo` is:
```
admin@windows-01 MINGW64 ~
$ which cargo
/c/Users/admin/scoop/shims/cargo
```
So I've installed it for the `admin` user, but not globally, so it's not in `/c/ProgramData/scoop/shims`, which is in `PATH`.

Here's the fix: https://github.com/status-im/infra-ci/commit/d7c52691
```
admin@windows-01 MINGW64 ~
$ which cargo
/c/ProgramData/scoop/shims/cargo
```
Sorry, I forgot `scoop` doesn't install globally by default when I made https://github.com/status-im/infra-ci/commit/5f191125.

**Comment by staheri14** [comment]
Blocked by #750

**Comment by staheri14** [comment]
Despite https://github.com/status-im/infra-ci/blob/ceb051d2d5c3ccca11d96277e171f6c70e493666/ansible/group_vars/ci-slave-windows.yml#L12 ,`npm` is still unknown on Jenkins windows machines ([logs](https://ci.status.im/blue/rest/organizations/jenkins/pipelines/nim-waku/pipelines/prs/pipelines/windows/branches/PR-786/runs/14/nodes/59/steps/65/log/?start=0)).

```
[2022-01-21T19:14:16.263Z] + make V=0 test2
[2022-01-21T19:14:16.263Z] /usr/bin/sh: npm: command not found
[2022-01-21T19:14:16.263Z] /usr/bin/sh: npx: command not found
```
 Any thought why it can be the case? thanks. @jakubgs

**Comment by staheri14** [comment]
> Despite https://github.com/status-im/infra-ci/blob/ceb051d2d5c3ccca11d96277e171f6c70e493666/ansible/group_vars/ci-slave-windows.yml#L12 ,`npm` is still unknown on Jenkins windows machines ([logs](https://ci.status.im/blue/rest/organizations/jenkins/pipelines/nim-waku/pipelines/prs/pipelines/windows/branches/PR-786/runs/14/nodes/59/steps/65/log/?start=0)).
> 
> ```
> [2022-01-21T19:14:16.263Z] + make V=0 test2
> [2022-01-21T19:14:16.263Z] /usr/bin/sh: npm: command not found
> [2022-01-21T19:14:16.263Z] /usr/bin/sh: npx: command not found
> ```
> 
> @jakubgs Any thought why it can be the case? thanks.

A gentle reminder re my previous comment @jakubgs Could you please verify whether `npm` is installed on Jenkins windows machines? this would help in debugging the issue. Thanks.

**Comment by jakubgs** [comment]
Sorry, I missed you message.

It's because the Node distribution from scoop does not create any shims for the global `bin` path:
https://github.com/ScoopInstaller/Main/blob/master/bucket/nodejs-lts.json

So they don't show up in `$PATH`. Also, for some reason binaries are not in the `bin` path:
```
admin@windows-02 MINGW64 ~
$ find /c/ProgramData/scoop/ -iname 'node.exe'
/c/ProgramData/scoop/apps/nodejs-lts/16.13.2/node.exe

admin@windows-02 MINGW64 ~
$ ls -l /c/ProgramData/scoop/apps/nodejs-lts/current/bin/
total 0

admin@windows-02 MINGW64 .../jenkins/agent
$ ls -l /c/ProgramData/scoop/apps/nodejs-lts/current/npm
-rwxr-xr-x 1 admin 197121 1409 Jan 10 17:22 /c/ProgramData/scoop/apps/nodejs-lts/current/npm*
```
This should do it: https://github.com/status-im/infra-ci/commit/64458908

Although I'm still confused as to why a Nim project needs Node.js tools to run tests... why is that?

**Comment by staheri14** [comment]
> Although I'm still confused as to why a Nim project needs Node.js tools to run tests... why is that?

npm/npx are required to install/run ganache-cli which is a blockchain simulator. Some of our tests (specifically waku-rln-relay tests) require interaction with the blockchain and for them we use ganache-cli.

**Comment by jakubgs** [comment]
That's awful, also looks like some prebuilt binaries that come with those packages have permissions which make it impossible for `jenkins`user to delete them after build due to `cleanWs()` call:

image

```
ERROR: Cannot delete workspace: Unable to delete 'C:\Users\jenkins\workspace\nim-waku_prs_windows_PR-786'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts. (Discarded 2 additional exceptions)
```
```
admin@windows-02 MINGW64 ~
$ ls -l /c/Users/jenkins/workspace/nim-waku_prs_windows_PR-786/node_modules/ganache-cli/node_modules/keccak/prebuilds/win32-x64/node.napi.node
-rwxr-xr-x 1 admin 197121 182784 Jan 25 03:40 /c/Users/jenkins/workspace/nim-waku_prs_windows_PR-786/node_modules/ganache-cli/node_modules/keccak/prebuilds/win32-x64/node.napi.node*
```
Not sure what I'm supposed to do about that. All builds will be marked as failed, or we have to drop `cleanWs()` on Windows.

UGH

**Comment by jakubgs** [comment]
WTF, is this, it does have permissions to write to the file, so why can't it remove it?

image

Maybe it's because it keep running even after the tests are done? Is that possible?

**Comment by staheri14** [comment]
> Maybe it's because it keep running even after the tests are done? Is that possible?

@jakubgs 
Yes, it keeps running afterward, we do have the kill command for macos when the tests are done https://github.com/status-im/nim-waku/blob/d45e3c8f7aa7c8c8eff85b6719672293834571aa/Makefile#L151, but not on windows.

**Comment by jakubgs** [comment]
Can't we just drop the conditional then?

**Comment by staheri14** [comment]
@jakubgs 
We can, but I am not sure whether the same command exists on windows, it is `pkill` which I think must have an equivalent like `taskkill` on widows.

**Comment by staheri14** [comment]
@jakubgs 
I am looking into how to kill a process on windows, in the meanwhile, would be also thankful to know your insights.

**Comment by staheri14** [comment]
With some browsing, found out we can kill all the node processes by `taskkill /im node.exe /F` though I think it is prudent to kill just ganache-cli

File: tests/v2/test_waku_rln_relay.nim [hunk 1]
```diff
@@ -136,125 +136,126 @@ proc uploadContract(ethClientAddress: string): Future[Address] {.async.} =
...
-      ethAccountAddress = accounts[9]
-    await web3.close()
-
-    # create an RLN instance
-    var rlnInstance = createRLNInstance()
-    check: rlnInstance.isOk == true
-
-    # generate the membership keys
-    let membershipKeyPair = membershipKeyGen(rlnInstance.value)
+#   asyncTest  "contract membership":
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:139]: Is this commented out on purpose? If so we should add a comment explaining why it is commented out, perhaps with a pointer to an issue describing the problem in more detail.

File: tests/v2/test_waku_rln_relay.nim [hunk 2]
```diff
@@ -136,125 +136,126 @@ proc uploadContract(ethClientAddress: string): Future[Address] {.async.} =
   return contractAddress
 
 procSuite "Waku rln relay":
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:138]:
```suggestion
procSuite "Waku rln relay":
# XXX Ganache tests commented out due to Ganache issue on Windows, see https://github.com/status-im/nim-waku/issues/833
```

File: Makefile [hunk 3]
```diff
@@ -133,14 +135,19 @@ else
...
+else ifeq ($(RLN), true) 
+	npm install ganache-cli; npx ganache-cli -p	8540	-g	0	-l	3000000000000&
+else  ifeq ($(CI), true)
 	npm install ganache-cli; npx ganache-cli -p	8540	-g	0	-l	3000000000000&
 endif
 
 rlnlib:
 ifeq ($(RLN), true)
 	cargo build --manifest-path vendor/rln/Cargo.toml
+else  ifeq ($(CI), true)
```
Comment (oskarth) [inline Makefile:149]: It seems a bit odd to me that we now have:

- RLN flag, off by default
- RLN is built in CI env
- But tests are commented out due to Windows Ganache issue

Wouldn't something like this make more sense:
- RLN flag is off by default
- Ganache tests are dealt with separate flag
- Ganache flag/behaviour is off for Windows CI (or all CI if it it isn't straightforward to detect OS (?))

Comment (staheri14) [reply Makefile:151]: Sure, I introduced a new flag `ONCHAIN_RLN` that controls rln-relay tests that need interaction with the `ganache-cli` process. Now we have:
- `RLN` and  `ONCHAIN_RLN` set off by default
- CIs: `RLN` is on and  `ONCHAIN_RLN` is off

**Comment by staheri14** [comment]
I also needed to address a flaky test  (epoch validation failure in `validateMessage test` test due to unpredictable delays in CIs) in this commit  https://github.com/status-im/nim-waku/pull/786/commits/9a78a10f10ab0ac0a10764c84760e5d27ee888b0

File: tests/v2/test_waku_rln_relay.nim (hunk 1 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Tests that need ganache-cli are now moved under the `onchain_rln` compiler flag. Please see the updated file.

**Comment by staheri14** [comment]
thanks @oskarth for the comments. 
I introduced a new flag `ONCHAIN_RLN` that controls rln-relay tests that need interaction with the `ganache-cli` process. Now we have:
- `RLN` and  `ONCHAIN_RLN` set off by default
- CIs: `RLN` is on and  `ONCHAIN_RLN` is off

**Comment by staheri14** [comment]
I also needed to address a flaky test  (epoch validation failure in `validateMessage test` test due to unpredictable delays in CIs) in this commit  https://github.com/status-im/nim-waku/pull/786/commits/9a78a10f10ab0ac0a10764c84760e5d27ee888b0

**Review by oskarth — APPROVED** [review]

**Review by oskarth — APPROVED** [review]

**Pull request merged**